### ISSUE TITLE
anyway to display link target?

### ISSUE THREAD

**Issue by ri-aje** [ranger/ranger#2036]
#### Runtime Environment

- Operating system and version: 
- Terminal emulator and version: 
- Python version: 
- Ranger version/commit: 
ranger version: ranger-master v1.9.3-27-gd17e532a
Python version: 3.7.1 (default, Sep 21 2019, 10:53:58) [GCC 7.4.0]
Locale: en_US.UTF-8
- Locale: 

#### Current Behavior
on the main column, soft link is displayed with an arrow -> to the right. but the linked target isn't shown. the only place showing the target seems to be the status line at the bottom, but there is a noticeable break as I have to move eye focus back and forth to the screen bottom. works pretty bad for large screen. 

#### Expected Behavior
would like the target shown after the ->, e.g., 'softlink                        -> target'. didn't find any flags to control this behavior, so I think it's missing.

#### Context

harder to find out what a link points to.

#### Possible Solutions
apart from rendering the -> sign, also append the link target?

#### Steps to reproduce
1. mkdir test && cd test
2. touch target
3. ln -s target link
4. launch ranger and observe the link line doesn't show target.

#### Traceback

```

```

**Comment by toonn** [comment]
Doesn't the preview tell you enough about the file a link points to?

**Comment by Rom1deTroyes** [comment]
I think the request is more about showing the targets in the listing area, and not only in the status bar ?

Make me think about the difference between `ls` and `ls -l`, and pressing [tab] for auto-completion :

```bash
❯ ls /lib/lib*
/lib/libhandle.so.1      /lib/libnih-dbus.so.1      /lib/libnih.so.1      /lib/libxtables.so.10
/lib/libhandle.so.1.0.3  /lib/libnih-dbus.so.1.0.0  /lib/libnih.so.1.0.0  /lib/libxtables.so.10.0.0

❯ ls /lib/lib  # pressing [Tab]                                                 
libhandle.so.1@       libnih-dbus.so.1@     libnih.so.1@          libxtables.so.10@                                             
libhandle.so.1.0.3    libnih-dbus.so.1.0.0  libnih.so.1.0.0       libxtables.so.10.0.0 

❯ ls -l /lib/lib*
lrwxrwxrwx 1 root root     18 févr. 22  2019 /lib/libhandle.so.1 -> libhandle.so.1.0.3
-rw-r--r-- 1 root root  14480 févr. 22  2019 /lib/libhandle.so.1.0.3
lrwxrwxrwx 1 root root     20 déc.   2  2018 /lib/libnih-dbus.so.1 -> libnih-dbus.so.1.0.0
-rw-r--r-- 1 root root  38920 déc.   2  2018 /lib/libnih-dbus.so.1.0.0
lrwxrwxrwx 1 root root     15 déc.   2  2018 /lib/libnih.so.1 -> libnih.so.1.0.0
-rw-r--r-- 1 root root 100376 déc.   2  2018 /lib/libnih.so.1.0.0
lrwxrwxrwx 1 root root     20 nov.   8  2014 /lib/libxtables.so.10 -> libxtables.so.10.0.0
-rw-r--r-- 1 root root  51896 nov.   8  2014 /lib/libxtables.so.10.0.0

```
Ranger has a mix between "showing which files are links" in the listing view, and "showing details about selected link" in the status bar :

![image](https://user-images.githubusercontent.com/8625432/86906043-3bcc3700-c113-11ea-941b-629370fc6e48.png)

The question may be : What is the cost of having the last column of the `ls -l` result in the liting view (in term of performence, readability, and so on) ?

**Comment by toonn** [comment]
I know OP is asking to include the link's target with the arrow next to the file name. It's a bit of a weird mixing of filename and contents and only works in very wide terminals but providing a setting wouldn't be out of the question.

My question still stands though, why do you need to know the target. Links are supposed to be treated as the file they point to, that's why we use the target's stat information and preview.

**Comment by ri-aje** [comment]
> I know OP is asking to include the link's target with the arrow next to the file name. It's a bit of a weird mixing of filename and contents and only works in very wide terminals but providing a setting wouldn't be out of the question.
> 
> My question still stands though, why do you need to know the target? Links are supposed to be treated as the file they point to, that's why we use the target's stat information and preview.

@toonn a couple of scenarios I hit that could benefit from showing the targets.
1. when I clean up folders mixed with links, I have cases where the folder names are very similar, e.g., `AA` vs. `A.A` and there are links pointing to them (links here function more like renaming the same thing or providing more convenient names to the same thing). in this case, I would like to quickly eyeball the whole picture like what is pointing to which and determine if I should merge `AA` into `A.A` or the other way around, or make a new one to absorb both of them, as well as cleaning up old links and make new ones. having a view show the targets (this is made even more powerful with filter) will make this very clear. preview doesn't help here, because it makes the links totally transparent, whereas in some cases, what it links to is the critical info. currently I have to look at the status line and remember the whole picture in my head. needless to say, that's inconvenient.
2. another case I ran into is a top folder containing many links to files in deep subfolders which categorizes those files based on some criteria, e.g., different versions, different times, etc, and the top folder essentially presents an overview to the whole tree (and this can get finer as each internal node may maintain additional links to its children, etc.). here the linkage structure is good info in itself (it's essentially showing the tree structure), and there is merit to show them, instead of treating links as transparent placeholders void of any other value.
3. the default `ls -l` does show link targets. I think it's a good design. it's useful for people who/when need them, and can be easily ignored if not needed, whereas the other way around doesn't offer this accommodation to both groups, i.e., if `ls -l` doesn't show targets by default, it gonna be painful to hunt them down, so maybe long proven unix design is another justification? 

that said, I totally appreciate ranger's current presentation of links and understand there are cases where this way works better (small screen or narrow column are good reasons), so all that I am looking for is a flag to turn on the feature (and I don't think it's too hard/inefficient to do it, right? for otherwise, how does `ls` do it?), and if you'd like do it better, maybe another hotkey to toggle the view. does this sound reasonable?

**Comment by toonn** [comment]
I suspect this is simple enough. Since the target's already shown in the statusbar it'd probably be easy to append it to the `->`  conditionally on a setting. I'm not sure how truncation would be handled when the column isn't wide enough though.

Another possibility is to be able to toggle the preview for links so the preview pane shows the target.

**Comment by ri-aje** [comment]
> I suspect this is simple enough. Since the target's already shown in the statusbar it'd probably be easy to append it to the `->` conditionally on a setting. I'm not sure how truncation would be handled when the column isn't wide enough though.

how does truncation work for file names?

> 
> Another possibility is to be able to toggle the preview for links so the preview pane shows the target.

no. that suffers the same problem of link preview now. the problem is I don't get any preview at all unless my cursor is on the line. whereas showing the target along side with the link gives a general view that presents all links and targets in the folder.

**Comment by ri-aje** [comment]
> > I suspect this is simple enough. Since the target's already shown in the statusbar it'd probably be easy to append it to the `->` conditionally on a setting. I'm not sure how truncation would be handled when the column isn't wide enough though.
> 
> how does truncation work for file names?
> 
> > Another possibility is to be able to toggle the preview for links so the preview pane shows the target.
> 
> no. that suffers the same problem of link preview now. the problem is I don't get any preview at all unless my cursor is on the line. whereas showing the target along side with the link gives a general view that presents all links and targets in the folder.

**Comment by ri-aje** [comment]
@toonn 
fyi, https://github.com/ri-aje/ranger/commit/01fc4bd3606b467755c03a9818e305cb32a9a073 works for me.

**Comment by ri-aje** [comment]
> @toonn
> fyi, [ri-aje@01fc4bd](https://github.com/ri-aje/ranger/commit/01fc4bd3606b467755c03a9818e305cb32a9a073) works for me.

hmm, not all the time. for larger folders, the target gets displayed only after my cursor passed through the link. looks like some additional updates were made when the link gets the cursor focus.

**Comment by mizhozan** [comment]
My whole project management and time management, prioritization, etc is filesystem/plain-text + ranger. I used symlinks a lot for e.g. prioritization when a file accessible from multiple directories. There are many cases when I need to see the target. In this cases I have three different scenario which only one might need a change:

1. When the symlink is a directory, it's easy I just traverse the symlink and `gl` to find real path. So no need for this functionality though I wish having undo sort of command which bring back previous path.
2. When the symlink is broken which ranger already support showing that. So no need for this.
3. When the symlink is a file and I like to eyeball the main parent directory it belongs to. Unlike case 1, I can not traverse its alternate directory and I have to spown shell and `ls -l`.

Here, I only need occasional use of `ls -l` for case 3. It would be nice to have a command like `gl` which show the target e.g. in statusbar instead.

**Comment by markus-bauer** [comment]
@mizhozan 

1. `H` goes back in history and `cd -` goes back to the previous dir. Both should do what you describe. If they don't, then something is broken.

3. Don't know if I understood this correctly. Instead of using gl, you can use gL on the link. This follows the highlighted link to the target.

**Comment by markus-bauer** [comment]
About the original request:

What doesn't work is:
1) Appending to the filetitle. Because truncating would be wrong.
2) Prepending to the infostring. Because infostrings would get too long and would be cut off most of the time.

I made a version that adds the link after the filetitle during drawing:
![2022-05-16_09-39](https://user-images.githubusercontent.com/23356070/168564476-e60b4617-f8bf-412c-b035-fcba993dfd75.png)


You can try it here:
https://github.com/markus-bauer/ranger/tree/show_links_in_browser
To enable it `set show_links_in_browser true`.
Or toggle by adding this to rc.conf `map zl set show_links_in_browser!`.


----
Notes:
- This uses the same code for resizing as the filetitle. So there should be correct truncating
( '-> filename_of_a_file.txt' shortens to '-> filen~.txt').
- It should dynamically switch from full link to '->' + infostring when the size gets too small (or the setting is off).
- The text for the link could have been simply the realpath, or use readlink like statusbar. But I made it a little more complex (see https://github.com/markus-bauer/ranger/blob/440347a8682be127dcf1efa2d451d79c4362548b/ranger/gui/widgets/browsercolumn.py#L543.)
- This uses pathlib and therefore requires python >= 3.4.
- The link is only added for drawing. It's not part of the entry name, and you can't use it for search or filter.
- Different coloring for the target would be nice to have, but not supported by this version. You would have to determine the filetype of the target.
- The target of the link is different than what the statusbar shows. Statusbar uses readlink, and shows the direct target, which could itself also include links. What I did is like readlink -f. It resolves the entire path, and will show the final target of the link. I think this is better, because this is also where `gL` will take you.

You can help test this:
1) When the setting is off, everything should be the same as now.
2) Does resizing work? It should truncate correctly and dynamically switch when the columns get too narrow.
3) What should the link show? readlink or readlink -f?
4) All existing linemodes and other line elements (like vcs) should just work.
5) Does everything else work? Have I broken anything?

If everything is okay, I can turn it into a PR.
